guide/src/python_from_rust.md

```diff
@@ -4,508 +4,48 @@ This chapter of the guide documents some ways to interact with Python code from
+- There is no concept of exclusive (`&mut`) references; any reference can mutate a Python object
 
-```rust
-use pyo3::prelude::*;
-use pyo3::types::PyTuple;
+PyO3's API reflects this by providing smart pointer types, `Py<T>`, `Bound<'py, T>`, and (the very rarely used) `Borrowed<'a, 'py, T>`. These smart pointers all use Python reference counting. See the [subchapter on types](./types.md) for more detail on these types.
 
-fn main() -> PyResult<()> {
-    let arg1 = "arg1";
-    let arg2 = "arg2";
-    let arg3 = "arg3";
+### Mutable access to `#[pyclass]` types
 
-    Python::with_gil(|py| {
-        let fun: Py<PyAny> = PyModule::from_code_bound(
-            py,
-            "def example(*args, **kwargs):
-                if args != ():
-                    print('called with args', args)
-                if kwargs != {}:
-                    print('called with kwargs', kwargs)
-                if args == () and kwargs == {}:
-                    print('called with no arguments')",
-            "",
-            "",
-        )?
-        .getattr("example")?
-        .into();
+The situation is helped by the Global Interpreter Lock (GIL), which ensures that only one thread can use the Python interpreter and its API at the same time, while non-Python operations (system calls and extension code) can unlock the GIL. (See [the section on parallelism](parallelism.md) for how to do that in PyO3.)
 
-        // call object without any arguments
-        fun.call0(py)?;
 
-        // call object with PyTuple
-        let args = PyTuple::new_bound(py, &[arg1, arg2, arg3]);
-        fun.call1(py, args)?;
 
-        // pass arguments as rust tuple
-        let args = (arg1, arg2, arg3);
-        fun.call1(py, args)?;
-        Ok(())
-    })
-}
-```
+The latter two points are the reason why some APIs in PyO3 require the `py:
+Python` argument, while others don't.
 
-### Creating keyword arguments
+The PyO3 API for Python objects is written such that instead of requiring a
+mutable Rust reference for mutating operations such as
+[`PyList::append`][PyList_append], a shared reference (which, in turn, can only
+be created through `Python<'_>` with a GIL lifetime) is sufficient.
 
-For the `call` and `call_method` APIs, `kwargs` can be `None` or `Some(&PyDict)`. You can use the [`IntoPyDict`]({{#PYO3_DOCS_URL}}/pyo3/types/trait.IntoPyDict.html) trait to convert other dict-like containers, e.g. `HashMap` or `BTreeMap`, as well as tuples with up to 10 elements and `Vec`s where each element is a two-element tuple.
+However, Rust structs wrapped as Python objects (called `pyclass` types) usually
+*do* need `&mut` access.  Due to the GIL, PyO3 *can* guarantee thread-safe access
+to them, but it cannot statically guarantee uniqueness of `&mut` references once
+an object's ownership has been passed to the Python interpreter, ensuring
+references is done at runtime using `PyCell`, a scheme very similar to
```

Comment:
Since we deprecated `PyCell` we should word this differently

Reply:
Agreed, I've removed this bit in the draft I'll push now.